Piwik collector

leelongmore opened the pull request:
Tests and code for a Piwik collector capable of supporting the fetching, parsing and pushing of realtime (see realtime.py module) and non-realtime (see query.py module) data from a Piwik server.

Example configuration files:

CREDENTIALS

{
  "token_auth": "anonymous",
  "url": "http://demo.piwik.org/"
}

REALTIME

{
  "data-set": {
    "data-group": "pay-foreign-marriage-certificates",
    "data-type": "realtime"
  },
  "entrypoint": "performanceplatform.collector.piwik.realtime",
  "query": {
    "site_id": "7",
    "api_method": "Live.getCounters",
    "minutes": "1"
  },
  "options": {},
  "token": "piwik"
}


NON-REALTIME

{
  "data-set": {
    "data-group": "pay-foreign-marriage-certificates",
    "data-type": "browser-count"
  },
  "entrypoint": "performanceplatform.collector.piwik.core",
  "query": {
    "site_id": "7",
    "api_method": "DevicesDetection.getBrowsers",
    "frequency": "weekly"
  },
  "options": {
    "mappings": {
      "label": "browser",
      "nb_visits": "sessions"
    },
    "idMapping": ["dataType","_timestamp", "timeSpan", "browser"]
  },
  "token": "piwik"
}

"frequency" is optional and defaults to weekly.

A hash can be added to "query" to support API method-specific params e.g.

  "query": {
    "site_id": "7",
    "api_method": "Goals.get",
    "api_method_arguments": {
       "idGoal": "1"
    }
  },

README files will be updated in performanceplatform-collector and performance platform-collector-config after successful testing of collector with Kainos on behalf of FCO.

timmow commented on performanceplatform/collector/piwik/base.py:
```diff
@@ -0,0 +1,29 @@
+from performanceplatform.utils import requests_with_backoff
```
gcloud, ga, and pingdom all have a core.py

webtrends has a base.py

given we are in an inconsistent state, i'm not sure how important this is, but if we went with majority rule this should be called core.py

timmow replied:
Actually looking at it, this is probably ok - it may be best to rename `query.py` to `core.py` to make more consistent with ga / pingdom - again webtrends is inconsistent here.

I'd like to know @tombooth and @jcbashdown s thoughts on this as well

timmow commented:
Its probably worth squashing some of these commits, like the tidy up and consistency ones

timmow commented on performanceplatform/collector/piwik/query.py:
```diff
@@ -0,0 +1,95 @@
+        self.date = Fetcher.set_piwik_date(start_at, end_at)
+        self.api_method_arguments = query.get('api_method_arguments', None)
+
+    @classmethod
```
I think these should be `@staticmethod` as they dont use the cls variable. Or, in python its fine to have methods that are not part of classes

timmow commented on performanceplatform/collector/piwik/query.py:
```diff
@@ -0,0 +1,95 @@
+            special_fields.append(mapped_fields)
+        return base_items, special_fields
+
+    @classmethod
```
again, '@staticmethod` or a standalone method would be better

timmow commented on tests/performanceplatform/collector/piwik/test_realtime.py:
```diff
@@ -0,0 +1,81 @@
+        return json.loads(f.read())
+
+
+def build_fetcher(query_params={}):
```
these methods could be in a utils file and imported into both tests to avoid duplication

leelongmore commented:
Made following changes in response to comments by @timmow:

- Replaced class methods with static methods
- Renamed query module to 'core'
- Squashed commits

We agreed not to change the test helper methods due to the different implementation of the Fetcher class in realtime.py and core.py.

timmow commented:
Sorry to be picky, but i'd remove the "prepare for merge" commit and squash it into the first. Its good to see it for review purposes though!

leelongmore commented:
Squashed the 'Prepare for merge' commit.
